addons/binding/org.openhab.binding.homematic/src/main/java/org/openhab/binding/homematic/internal/discovery/CcuDiscoveryService.java

```diff
@@ -0,0 +1,142 @@
+ * are made available under the terms of the Eclipse Public License v1.0
+ * which accompanies this distribution, and is available at
+ * http://www.eclipse.org/legal/epl-v10.html
+ */
+package org.openhab.binding.homematic.internal.discovery;
+
+import static org.openhab.binding.homematic.HomematicBindingConstants.THING_TYPE_BRIDGE;
+
+import java.io.IOException;
+import java.net.DatagramPacket;
+import java.net.InetAddress;
+import java.net.MulticastSocket;
+import java.net.SocketTimeoutException;
+import java.util.Collections;
+import java.util.concurrent.Future;
+
+import org.eclipse.smarthome.config.discovery.AbstractDiscoveryService;
+import org.eclipse.smarthome.config.discovery.DiscoveryResultBuilder;
+import org.eclipse.smarthome.config.discovery.DiscoveryService;
+import org.eclipse.smarthome.core.thing.ThingUID;
+import org.openhab.binding.homematic.internal.discovery.eq3udp.Eq3UdpRequest;
+import org.openhab.binding.homematic.internal.discovery.eq3udp.Eq3UdpResponse;
+import org.osgi.service.component.annotations.Component;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * Discovers Homematic CCU's and adds the results to the inbox.
+ *
+ * @author Gerhard Riegler - Initial contribution
+ */
+@Component(immediate = true, service = DiscoveryService.class)
+public class CcuDiscoveryService extends AbstractDiscoveryService {
+    private final Logger logger = LoggerFactory.getLogger(CcuDiscoveryService.class);
+
+    private static final int RECEIVE_TIMEOUT = 3000;
+    private InetAddress broadcastAddress;
+    private MulticastSocket socket;
+    private Future<?> scanFuture;
+
+    public CcuDiscoveryService() {
+        super(Collections.singleton(THING_TYPE_BRIDGE), 5, true);
+    }
+
+    @Override
+    protected void startScan() {
+        if (scanFuture == null) {
+            scanFuture = scheduler.submit(this::startDiscovery);
+        } else {
+            logger.debug("Homematic CCU background discovery scan in progress");
+        }
+    }
+
+    @Override
+    protected void stopScan() {
+        if (scanFuture != null) {
+            scanFuture.cancel(false);
+            scanFuture = null;
+        }
+    }
+
+    @Override
+    protected void startBackgroundDiscovery() {
+        // only start once at startup
+        startScan();
+    }
+
+    @Override
+    protected void stopBackgroundDiscovery() {
+        stopScan();
+    }
+
+    private synchronized void startDiscovery() {
+        try {
+            logger.debug("Starting Homematic CCU discovery scan");
+            broadcastAddress = InetAddress.getByName("255.255.255.255");
+            socket = new MulticastSocket();
+            socket.setBroadcast(true);
+            socket.setTimeToLive(5);
+            socket.setSoTimeout(800);
+
+            sendBroadcast();
+            receiveResponses();
+        } catch (Exception ex) {
+            logger.error("{}", ex.getMessage(), ex);
```

Comment:
I would prefer the logging to have some context, I believe it could potentially only return the classname and our configuration does not return a line number anymore so a message that "An error was thrown while running discovery" would imho not be misplaced.